### DISCUSSION TITLE
Does Stockfish pick a move based on the highest centipawns score?

### DISCUSSION THREAD

**Discussion by MatthewDulchinos** [official-stockfish/Stockfish#3810] [Q&A]
I am trying to write a program to evaluate my moves, using Stockfish as a way to judge my moves. However, during some of the moves I make, Stockfish recommends a move that results in a lower score than the move I made. Checking the same game on Lichess or Chess.com recommends the same move that Stockfish makes, but I am confused on why my move is rated as a "better" move. 

Does Stockfish not make moves based on the highest possible CP score? If not, how do chess programs identify "blunders"?

**Comment by niklasf** [comment]
Yes, Stockfish selects the recommended move according to the score.

For evaluations, Stockfish communicates only centipawn scores and optionally expected chances for winning/drawing/losing. A GUI like Chess.com needs to implement its own heuristics or cutoffs for score differences, if it wants to classify moves as "blunders" or similar.

The evaluation given for a position can definitely differ from the evaluation of the position after the recommended move, in either direction.

Imagine this would not be allowed. Suppose you have two chess engines playing against each other, so the entire game consists of recommended engine moves. The evaluation of the final position can only be *win*, *draw* or *loss*. Then the position before that must also have the same evaluation, and so on, until the starting position is classified as *win* or *draw* or *loss*. And chess is solved.

Obviously that's not what happens. Chess engines are not perfect and each position is analysed with limited computing time only. So playing one more move, even the recommended move, and then continuing the search, usually leads to more precise evaluations of the resulting position. For example, the continued analysis may benefit from the previous state of the hashtable, and more importantly it does not have to spend time on analyzing the other candidate moves.
